pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,364 @@
+    Of these arguments, all elements are included in ``encoding_args``
+    """
+    sig_pars = signature(qnode.func).parameters
+    arg_names = list(sig_pars.keys())
+    arg_names_no_def = [name for name, par in sig_pars.items() if par.default is par.empty]
+
+    if encoding_args is None:
+        if argnum is None:
+            encoding_args = OrderedDict((name, ...) for name in arg_names_no_def)
+            argnum = list(range(len(arg_names_no_def)))
+        elif np.isscalar(argnum):
+            encoding_args = OrderedDict({arg_names[argnum]: ...})
+            argnum = [argnum]
+        else:
+            encoding_args = OrderedDict((arg_names[num], ...) for num in argnum)
+    else:
+        requested_names = set(encoding_args)
+        if not all(name in arg_names for name in requested_names):
+            raise ValueError(
+                f"Not all names in {requested_names} are known. " f"Known arguments: {arg_names}"
+            )
+        # Selection of requested argument names from sorted names
+        if isinstance(encoding_args, set):
+            encoding_args = OrderedDict(
+                (name, ...) for name in arg_names if name in requested_names
+            )
+        else:
+            encoding_args = OrderedDict(
+                (name, encoding_args[name]) for name in arg_names if name in requested_names
+            )
+        argnum = [arg_names.index(name) for name in encoding_args]
+
+    return encoding_args, argnum
+
+
+def expand_multi_par(tape, depth=10):
+    """Expand a tape until it does not contain any multi-parameter gates, if possible.
+
+    Args:
+        tape (.QuantumTape): Tape to be expanded
+        depth (int): Maximum expansion depth
+
+    Returns
+        .QuantumTape: Expanded tape
+
+    """
+    if any(len(op.parameters) > 1 for op in tape.operations):
+        stopping_cond = lambda g: (
+            isinstance(g, qml.measure.MeasurementProcess) or len(g.parameters) <= 1
+        )
+        new_tape = tape.expand(depth=depth, stop_at=stopping_cond)
+        params = new_tape.get_parameters(trainable_only=False)
+        new_tape.trainable_params = qml.math.get_trainable_indices(params)
+
+        return new_tape
+
+    return tape
+
+
+def advanced_spectrum(qnode, encoding_args=None, argnum=None, decimals=5, validation_kwargs=None):
+    r"""Compute the frequency spectrum of the Fourier representation of quantum circuits,
+    including classical preprocessing.
+
+    The circuit must only use single-parameter gates of the form :math:`e^{-i x_j G}` as
+    input-encoding gates, which allows the computation of the spectrum by inspecting the gates'
+    generators :math:`G`. The most important example of such gates are Pauli rotations.
+
+    .. note::
+
+        More precisely, the ``advanced_spectrum`` function relies on the gate to
+        define a ``generator``, and will fail if gates marked controlled by marked
+        parameters do not have this attribute.
```

Comment:
Check phrasing here, I didn't quite understand the end of the sentence.

Reply:
I think this notebox can be deleted given multi-parameter gates with no generator are now supported by this function

Reply:
Ah! So we should definitely expand operations that do not have a generator, right? :o

Reply:
oh, maybe I misunderstand, I thought that was the case?

Reply:
No, only multi-parameter gates were being expanded, based on `len(op.parameters)>1`. I now added
`not hasattr(op, "generator") or op.generator[0] is None`.